DDOT_module/child_classes/ddot_marginal_analyzer_vanilla.py

```diff
@@ -0,0 +1,396 @@
+            noise_matrix[:, int(input_state, 2)] = probability_vector[:, 0]
+        return noise_matrix
+
+    @staticmethod
+    def average_noise_matrices_over_some_qubits(matrices_cluster: Dict[str, np.ndarray],
+                                                all_neighbors: List[int],
+                                                qubits_to_be_left: List[int]) -> Dict[str, np.ndarray]:
+        """
+          Given dictionary of noise matrices, average them over some qubits.
+
+         :param matrices_cluster: dictionary for which KEY is classical INPUT state of neighbors,
+                                  and VALUE is a noise matrix
+         :param all_neighbors: list of neighbors of given cluster
+         :param qubits_to_be_left: qubits which we are interested in and we do not average over them
+
+         :return: dictionary of noise matrices
+                    depending on the state of neighbors MINUS qubits_to_be_averaged_over
+
+         """
+
+        reversed_enumerated = anf.get_reversed_enumerated_from_indices(all_neighbors)
+        averaging_normalization = int(2 ** (len(all_neighbors) - len(qubits_to_be_left)))
+
+        states_after_averaging = anf.register_names_qubits(range(len(qubits_to_be_left)),
+                                                           len(qubits_to_be_left), False)
+        averaged_dimension = list(matrices_cluster.values())[0].shape[0]
+
+        averaged_matrices_cluster = {
+            state: np.zeros((averaged_dimension, averaged_dimension), dtype=float) for state in
+            states_after_averaging}
+
+        qubits_to_be_averaged_over = list(set(all_neighbors).difference(set(qubits_to_be_left)))
+        qubits_to_be_averaged_over_mapped = [reversed_enumerated[q] for q in
+                                             qubits_to_be_averaged_over]
+
+        for neighbors_state, conditional_noise_matrix in matrices_cluster.items():
+            list_string_neighbors = list(copy.deepcopy(neighbors_state))
+
+            list_string_neighbors = list(np.delete(list_string_neighbors,
+                                                   qubits_to_be_averaged_over_mapped))
+
+            string_neighbors = ''.join(list_string_neighbors)
+
+            # TODO FBM: make sure this commented check is not needed anymore
+            # if string_neighbors == '' or string_neighbors:
+            #     try:
+            #         averaged_matrices_cluster['averaged'] +=
+            #         conditional_noise_matrix / averaging_normalization
+            #     except(KeyError):
+            #         averaged_matrices_cluster['averaged'] =
+            #         conditional_noise_matrix/averaging_normalization
+            # else:
+            averaged_matrices_cluster[
+                string_neighbors] += conditional_noise_matrix / averaging_normalization
+
+        return averaged_matrices_cluster
+
+    def _compute_noise_matrix_averaged(self,
```

Comment:
This method is quite large. Can you maybe decompose it into several smaller ones?

Reply:
I don't remember how it was back then, but now I think it's not too large